### PR TITLE
Remove cart_id from civicrm_participant, create new table in event cart

### PR DESCRIPTION
**Pull request by eileenmcnaughton**
Overview
----------------------------------------
Remove cart_id from civicrm_participant, create new table in event cart

Before
----------------------------------------
Participant field cart_id exists on civicm_participant just for event cart - which is not a recommended way to integrate

After
----------------------------------------
New extension specific table for storing participant_id->cart_id

Some r-run screenshots
image

'your cart' is a link here

image

image

some notices fixed - this one is too hard, double email is pre-existing
image

image

image

notices do not seem regressive

image

Technical Details
----------------------------------------
Event cart is on it's way out but has created ongoing complexity by it's half-in-half-out relationship to core - this finishes moving it's data structure to the extension

Comments
----------------------------------------

### CODE REVIEW

File: CRM/Upgrade/Incremental/php/FiveSeventySix.php [hunk 1]
```diff
@@ -43,6 +43,9 @@ public function upgrade_5_76_alpha1($rev): void {
     if (!CRM_Core_DAO::singleValueQuery('SELECT id FROM civicrm_extension WHERE full_name = "eventcart"')) {
       $this->addTask('Remove data related to disabled even cart extension', 'removeEventCartAssets');
     }
+    else {
+      $this->addTask('Migrate event cart ID', 'migrateEventCartID');
```
Comment (colemanw) [inline CRM/Upgrade/Incremental/php/FiveSeventySix.php:47]: Where is this task defined?

Comment (eileenmcnaughton) [reply CRM/Upgrade/Incremental/php/FiveSeventySix.php:48]: not you see it ....

File: ext/eventcart/CRM/Event/Cart/BAO/MerParticipant.php [hunk 2]
```diff
@@ -53,10 +57,18 @@ public static function create(&$params) {
       'status_id' => self::get_pending_in_cart_status_id(),
       'contact_id' => $params['contact_id'],
       'event_id' => $params['event_id'],
-      'cart_id' => $params['cart_id'],
     ];
     $participant = CRM_Event_BAO_Participant::create($participantParams);
-    $mer_participant = new CRM_Event_Cart_BAO_MerParticipant($participant);
+    if (!empty($params['cart_id']) && !EventCartParticipant::get(FALSE)
```
Comment (seamuslee001) [inline ext/eventcart/CRM/Event/Cart/BAO/MerParticipant.php:62]: @eileenmcnaughton why not use the save action instead?

Comment (eileenmcnaughton) [reply ext/eventcart/CRM/Event/Cart/BAO/MerParticipant.php:63]: @seamuslee001 cos I would need to know the ID to do a do an update I think - which = doing a get first

Comment (seamuslee001) [reply ext/eventcart/CRM/Event/Cart/BAO/MerParticipant.php:63]: I would have thought you could do this right?

```
EventCartParticipant::save(FALSE)
   ->setValues(['cart_id' => $params['cart_id'], 'participant_id' => $participant->id])
   ->setMatch(['cart_id', 'participant_id'])
  ->execute();
```

Comment (eileenmcnaughton) [reply ext/eventcart/CRM/Event/Cart/BAO/MerParticipant.php:63]: oh yeah that might work

Comment (eileenmcnaughton) [reply ext/eventcart/CRM/Event/Cart/BAO/MerParticipant.php:63]: (I didn't know about `setMatch()`

**Comment by eileenmcnaughton** [comment]
@colemanw this is close but I'm still missing something to make the auto-table generate work

**Comment by colemanw** [comment]
@eileenmcnaughton instead of running the file `CRM/Upgrade/Incremental/schema/5.76.alpha1.EventCartParticipant.entityType.php`, what about just running the extension installer. We have a helper for this called `addExtensionTask()`. I think the only gotcha there is that you need to set the weight of that task to run *before* the task which populates the new table.

**Comment by eileenmcnaughton** [comment]
@colemanw that function seems to be specifically for enabling extensions - this is already enabled - & is pretty edge/ wonky 
I think I'm maybe go with just running a `CREATE TABLE IF NOT Exists`

**Comment by eileenmcnaughton** [comment]
Oh - hang on - the table is not installed on new install either - for unknown reasons

**Comment by eileenmcnaughton** [comment]
OK - for known reasons - I was looking at the wrong DB

**Comment by eileenmcnaughton** [comment]
@colemanw well that worked anyway

File: ext/eventcart/CRM/Event/Cart/BAO/MerParticipant.php [hunk 3]
```diff
@@ -9,6 +9,8 @@
  +--------------------------------------------------------------------+
  */
 
+use Civi\Api4\EventCartParticipant;
+
 /**
  * Class CRM_Event_Cart_BAO_MerParticipant
```
Comment (colemanw) [inline ext/eventcart/CRM/Event/Cart/BAO/MerParticipant.php:15]: Are MerParticipants... when mermaids register for your event?

Comment (eileenmcnaughton) [reply ext/eventcart/CRM/Event/Cart/BAO/MerParticipant.php:15]: it's gender neutral

**Pull request merged**